### PR TITLE
enh(SQL): enhance sql redo sql query

### PR DESCRIPTION
**Pull request by asalhi-ext**
enhance sql redo sql query and fix service filter

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x
- [ ] 24.04.x
- [x] master

<h2> How this pull request can be tested ? </h2>

Check [Ticket](https://centreon.atlassian.net/browse/MON-147921)

#### Community contributors & Centreon team

- [ ] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/www/include/eventLogs/xml/data.php [hunk 1]
```diff
@@ -329,11 +322,21 @@
...
+            $HostCache[$h["host_name"]] = $h["host_address"];
+        }
+        $pearDB->closeQuery($dbResult);
+    } catch (CentreonDbException $e) {
+        CentreonLog::create()->error(
+            CentreonLog::TYPE_BUSINESS_LOG,
+            'Error while fetching hosts',
+            [],
+            $e
+        );
```
[the remark this answers is not in the document]
Comment (asalhi-ext) [reply centreon/www/include/eventLogs/xml/data.php:338]: resolved

File: centreon/www/include/eventLogs/xml/data.php [hunk 2]
```diff
@@ -291,29 +262,51 @@
...
-    $start = mktime($matchesT[1], $matchesT[2], "0", $matchesD[1], $matchesD[2], $matchesD[3]);
+    $dateTime = DateTime::createFromFormat('m/d/Y H:i', "$StartDate $StartTime");
+    if ($dateTime !== false) {
+        $start = $dateTime->getTimestamp();
+    } else {
+        $centreonLog::create()->error(
+            CentreonLog::TYPE_BUSINESS_LOG,
+            "Invalid date format: $StartDate $StartTime"
+        );
+    }
```
[the remark this answers is not in the document]
Comment (asalhi-ext) [reply centreon/www/include/eventLogs/xml/data.php:273]: resolved

File: centreon/www/include/eventLogs/xml/data.php [hunk 3]
```diff
@@ -361,506 +364,470 @@
...
+    $selectClause .= "DISTINCT ";
+}
 
-    $stmt = $pearDBO->prepare($req . $limitReq);
-    foreach ($queryValues as $bindId => $bindData) {
-        foreach ($bindData as $bindType => $bindValue) {
-            $stmt->bindValue($bindId, $bindValue, $bindType);
+// Add SQL_CALC_FOUND_ROWS
+$selectClause .= "SQL_CALC_FOUND_ROWS ";
+
```
[the remark this answers is not in the document]
Comment (asalhi-ext) [reply centreon/www/include/eventLogs/xml/data.php:590]: done

File: centreon/www/include/eventLogs/xml/data.php [hunk 4]
```diff
@@ -361,506 +364,470 @@
...
+            $meta = $pearDB->fetch($statement);
+            $pearDB->closeQuery($statement);
+        } catch (CentreonDbException $e) {
+            CentreonLog::create()->error(
+                CentreonLog::TYPE_BUSINESS_LOG,
+                'Error while fetching meta_services',
+                [],
+                $e
             );
         }
```
[the remark this answers is not in the document]
Comment (asalhi-ext) [reply centreon/www/include/eventLogs/xml/data.php]: done

File: centreon/www/include/eventLogs/xml/data.php [hunk 5]
```diff
@@ -46,10 +46,16 @@
...
 require_once __DIR__ . '/PaginationRenderer.php';
+require_once __DIR__ . '/../../../class/centreonDB.class.php';
+require_once __DIR__ . '/../../../class/centreonLog.class.php';
+require_once __DIR__ . '/../../../class/exceptions/CentreonDbException.php';
 
 // Connect to DB
-$pearDB = $dependencyInjector['configuration_db'];
-$pearDBO = $dependencyInjector['realtime_db'];
+$pearDB = new CentreonDB(CentreonDB::LABEL_DB_CONFIGURATION);
+$pearDBO = new CentreonDB(CentreonDB::LABEL_DB_REALTIME);
```
Comment (callapa) [inline centreon/www/include/eventLogs/xml/data.php:55]: Each time you create a new instance of this class, you create a new database connection.
To save on the number of open connections, it's preferable to use an existing connection, and therefore to use those already available in the $dependencyInjector variable.

**Review by callapa — CHANGES_REQUESTED** [review]

File: centreon/www/include/eventLogs/xml/data.php [hunk 6]
```diff
@@ -233,53 +209,48 @@
...
-    $unknown = "false";
-    $unreachable = "false";
-    $down = "false";
-    $warning = "false";
-    $critical = "false";
-    $acknowledgement = "false";
-    $oh = "false";
-    $notification = "false";
-    $alert = "false";
+    $ok = $up = $unknown = $unreachable = $down = $warning = $critical = $acknowledgement = $oh = $alert = "false";
```
Comment (callapa) [inline centreon/www/include/eventLogs/xml/data.php:253]: no clearer than before.

File: centreon/www/include/eventLogs/xml/data.php [hunk 7]
```diff
@@ -291,29 +262,51 @@
...
+        );
+    }
 }
 
-// setting the startDate/Time using the user's chosen period
-// and checking if the start date/time was set by the user, to avoid to display/export the whole data since 1/1/1970
+// Setting the startDate/Time using the user's chosen period
 $period = 86400;
 if ($auto_period > 0 || $start === 0) {
     $period = (int)$auto_period;
```
Comment (callapa) [inline centreon/www/include/eventLogs/xml/data.php:290]: This variable is already of type Integer
```suggestion
    $period = $auto_period;
```

File: centreon/www/include/eventLogs/xml/data.php (hunk 5 above)
Comment (asalhi-ext) [reply centreon/www/include/eventLogs/xml/data.php]: done

File: centreon/www/include/eventLogs/xml/data.php (hunk 6 above)
Comment (asalhi-ext) [reply centreon/www/include/eventLogs/xml/data.php]: resolved

File: centreon/www/include/eventLogs/xml/data.php [hunk 8]
```diff
@@ -107,99 +113,69 @@
...
+    'critical' => ['default' => 'true'],
+    'unknown' => ['default' => 'true'],
+    'acknowledgement' => ['default' => 'true'],
+    'notification' => ['default' => 'false'],
+    'alert' => ['default' => 'true'],
+    'oh' => ['default' => 'false'],
+    'error' => ['default' => 'false'],
+    'output' => ['sanitize' => true, 'default' => ''],
+    'search_H' => ['default' => 'VIDE'],
+    'search_S' => ['default' => 'VIDE'],
```
[the remark this answers is not in the document]
Comment (asalhi-ext) [reply centreon/www/include/eventLogs/xml/data.php:180]: its in the old code

File: centreon/www/include/eventLogs/xml/data.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (asalhi-ext) [reply centreon/www/include/eventLogs/xml/data.php]: done

File: centreon/www/include/eventLogs/xml/data.php [hunk 9]
```diff
@@ -291,29 +286,51 @@
...
-    $end = mktime($matchesT[1], $matchesT[2], "0", $matchesD[1], $matchesD[2], $matchesD[3]);
+    $dateTime = DateTime::createFromFormat('m/d/Y H:i', "$EndDate $EndTime");
+    if ($dateTime !== false) {
+        $end = $dateTime->getTimestamp();
+    } else {
+        $centreonLog->error(
+            CentreonLog::TYPE_BUSINESS_LOG,
+            "Invalid date format: $EndDate $EndTime"
+        );
+    }
```
[the remark this answers is not in the document]
Comment (asalhi-ext) [reply centreon/www/include/eventLogs/xml/data.php:308]: done

File: centreon/www/include/eventLogs/xml/data.php [hunk 10]
```diff
@@ -361,506 +363,494 @@
...
-    $stmt->closeCursor();
+    $logs = $pearDBO->fetchAll($statement);
+    $pearDBO->closeQuery($statement);
+} catch (CentreonDbException $e) {
+    CentreonLog::create()->error(
+        CentreonLog::TYPE_BUSINESS_LOG,
+        'Error while fetching logs',
+        $e->getOptions()
+    );
+}
```
[the remark this answers is not in the document]
Comment (asalhi-ext) [reply centreon/www/include/eventLogs/xml/data.php:708]: its only fetching data

File: centreon/src/Core/Common/Infrastructure/Repository/SqlMultipleBindTrait.php [hunk 11]
```diff
@@ -28,16 +28,17 @@ trait SqlMultipleBindTrait
...
      */
-    protected function createMultipleBindQuery(array $list, string $prefix): array
+    protected function createMultipleBindQuery(array $list, string $prefix, ?int $bindType = null): array
     {
         $bindValues = [];
         $list = array_values($list); // Reset index to avoid SQL injection
 
         foreach ($list as $index => $id) {
-            $bindValues[$prefix . $index] = $id;
+            $bindValues[$prefix . $index] = ($bindType === null) ? $id : [$id, $bindType];
```
Comment (callapa) [inline centreon/src/Core/Common/Infrastructure/Repository/SqlMultipleBindTrait.php:41]: You must comment out the new return values when the optional parameter is sent.

File: centreon/www/include/common/sqlCommonFunction.php [hunk 12]
```diff
@@ -28,15 +28,16 @@
 /**
  * @param array<int|string, int|string> $list
  * @param string $prefix
+ * @param int $bindType (optional)
  *
  * @return array{0: array<string, mixed>, 1: string}
  */
-function createMultipleBindQuery(array $list, string $prefix): array
+function createMultipleBindQuery(array $list, string $prefix, int $bindType = null): array
```
Comment (callapa) [inline centreon/www/include/common/sqlCommonFunction.php:35]: You must comment out the new return values when the optional parameter is sent.

File: centreon/src/Core/Common/Infrastructure/Repository/SqlMultipleBindTrait.php (hunk 11 above)
Comment (asalhi-ext) [reply centreon/src/Core/Common/Infrastructure/Repository/SqlMultipleBindTrait.php:44]: resolved

**Review by callapa — APPROVED** [review]

**Pull request merged**